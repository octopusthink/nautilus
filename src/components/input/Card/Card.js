import PropTypes from 'prop-types';
import React from 'react';
import styled, { css } from 'styled-components';

import spacerSizes from '../../../base/spacing';
import { Heading } from '../../typography/Heading';
import { Paragraph } from '../../typography/Paragraph';

export const cardOrientations = {
  portrait: css`
   grid-template-rows: auto auto;
  `,
  landscape: css`
    grid-template-columns: 1fr 2fr;
    align-items: center;
  `,
};

export const StyledCard = styled.article`
  cursor: pointer;
  display: grid;
  margin: 0 0 ${spacerSizes.large};
  box-shadow: rgba(0,0,0,0.05) 0 2px 16px
  ${props => cardOrientations[props.orientation]};
`;

export const StyledCardMedia = styled.div`
  margin: 0;
  line-height: 0;
`;

export const StyledImage = styled.img`
  max-width: 100%;
  object-fit: cover;
  width: 100%;
  height: 100%;
`;

export const StyledCardContent = styled.div`
  padding: ${spacerSizes.small};  
`;

export const Card = ({ children, onClick, orientation, title, metadata, thumbnail, media }) => (
  <StyledCard
    onClick={onClick}
    orientation={orientation}
    title={title}
    metadata={metadata}
    thumbnail={thumbnail}
    media={media}
  >

    {media && (
      <StyledCardMedia>
        <StyledImage src={media} />
      </StyledCardMedia>
    ) }

    <StyledCardContent>
      <Heading level={0} size="xxsmall">{metadata}</Heading>
      <Heading level={2} size="large">{title}</Heading>
      {children && (
        <Paragraph noMargin size="small">{children}</Paragraph>
      ) }
    </StyledCardContent>
  </StyledCard>
);

export const CardOrientations = Object.keys(cardOrientations);

Card.propTypes = {
  /** Content to be rendered inside a Card */
  children: PropTypes.node,
  /** Function to run on click */
  onClick: PropTypes.func,
  /** Orientation. Options are "vertical" or "horizontal". */
  orientation: PropTypes.oneOf(CardOrientations),
  /** Title and metadata are strings. */
  title: PropTypes.string,
  metadata: PropTypes.string,
  /** Thumbnail and media are React components, passed in as images/video/SVG/etc.` */
  thumbnail: PropTypes.element,
  media: PropTypes.string,
};

Card.defaultProps = {
  onClick: () => {},
  orientation: '',
};

export default Card;
