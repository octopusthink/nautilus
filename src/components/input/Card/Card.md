### Card

A card is used to represent a single, descrete piece of data. 

```js
<Card title="Cards can be super simple" metadata="Places I like to visit" />
```

```js
<Card orientation="landscape" title="Fish &  Chips & Vinegar" metadata="Places I like to visit" media="http://placekitten.com/400/800">
	<Paragraph size="small">By default, cards use a landscape orientation, but can also be swapped around. All content is optional.</Paragraph>
</Card>
```

```js
<Card orientation="portrait" title="A card can go both ways" metadata="Horizontal, or vertical?" media="http://placekitten.com/800/400">
	<Paragraph size="medium">An experimental project designed to vastly simplify and streamline the WordPress site building process for small business owners, by using technology to automate visual design decisions.</Paragraph>
</Card>
```

