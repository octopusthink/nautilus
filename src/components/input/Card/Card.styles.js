import styled, { css } from 'styled-components';

import fontSizes from '../../../base/typography/fontSizes';

export const cardOrientations = {
  horizontal: css`
   border-color: pink;
   flex-direction: column;
  `,
  vertical: css`
    border-color: blue;
    flex-direction: row;
  `,
};

export const StyledCard = styled.article`
  ${props => cardOrientations[props.orientation]};
  cursor: pointer;
  display: flex;
  margin: 0 0 0 1.6rem;
  box-shadow: rgba(0, 0, 0, 0.25) 0 0 2px;
`;

export const StyledCardMedia = styled.div`
  ${props => cardOrientations[props.orientation]};
  width: 100%;
  height: 50%;
  background: pink;
`;

export const StyledCardContent = styled.div`
  padding: 1.6rem;  
`;
