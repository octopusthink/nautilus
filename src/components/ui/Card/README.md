A card is used to represent a descrete and self-contained piece of data, typically as part of a feed or series of related datum.

```jsx
import { Paragraph, Button, Heading, Tags, Icon } from '@octopusthink/nautilus';

<React.Fragment>
  <Card title="Cards can be super simple">
    <Icon name="star" />
    <Heading level={1}>Cards should always have a title</Heading>
    <Paragraph small>The Card can either have a title attribute, or you can pass it a heading. Not sure which is the best approach here.</Paragraph>
    <Button>Choose this</Button>
    <Button minimal navigation>Learn more</Button>
  </Card>

  <Card title="Cards can be super simple">
    <Heading level={2}>Cards should always have a title</Heading>
    <Paragraph>The Card can either have a title attribute, or you can pass it a heading. Not sure which is the best approach here.</Paragraph>
  </Card>

  <Card title="Cards can be super simple">
    <Heading level={3}>Cards should always have a title</Heading>
    <Paragraph large>The Card can either have a title attribute, or you can pass it a heading. Not sure which is the best approach here.</Paragraph>
  </Card>
</React.Fragment>
```

```jsx
import { Paragraph } from '@octopusthink/nautilus';

<Card
  orientation="landscape"
  title="Fish & Chips & Vinegar"
  metadata="Places I like to visit"
  media="http://placekitten.com/400/800"
>
  <Paragraph size="small">
    By default, cards use a landscape orientation, but can also be swapped around.
    All content is optional.
  </Paragraph>
</Card>
```

```jsx
import { Paragraph } from '@octopusthink/nautilus';

<Card
  orientation="portrait"
  title="A card can go both ways"
  metadata="Horizontal, or vertical?"
  media="http://placekitten.com/800/400"
>
  <Paragraph size="medium">
    Dis convallis taciti molestie cum etiam mollis facilisis ligula ultricies, accumsan mattis phasellus malesuada in sollicitudin pulvinar quis, turpis pretium purus mi enim suspendisse nisl dolor. Egestas diam mollis bibendum integer est volutpat, quisque posuere cras eu. Eros enim maecenas tellus semper torquent platea pulvinar vulputate fermentum laoreet consequat vel purus, vitae rhoncus est sollicitudin lacus pretium id montes blandit phasellus nullam cras.
  </Paragraph>
</Card>
```

```jsx
import { Paragraph, Button, Heading, Tags, Icon } from '@octopusthink/nautilus';

<Card orientation="landscape">
  {/*
  <Card.Header>
    <img src="http://placekitten.com/400/800" />
    <Heading>Fish & Chips & Vinegar</Heading>
    <Tags><Tags.Tag>Places I like to visit</Tags.Tag></Tags>
  </Card.Header>
  */}

  <Paragraph size="small">
    By default, cards use a landscape orientation, but can also be
    swapped around. All content is optional.
  </Paragraph>

  {/*
  <Card.Footer>
    <Button>Learn more</Button>
    <Button minimal>Hide this</Button>
  </Card.Footer>
  */}
</Card>
```
