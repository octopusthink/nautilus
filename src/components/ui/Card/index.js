import { css } from '@emotion/core';
import styled from '@emotion/styled';
import PropTypes from 'prop-types';
import React, { forwardRef } from 'react';

import { headingSmall, toUnits } from 'styles';
import { useTheme } from 'themes';

export const Card = (props) => {
  const { children, ...otherProps } = props;
  const theme = useTheme();

  return (
    <div
      css={css`
        background: white;
        border: 1px solid ${theme.colors.neutral.grey200};
        box-shadow: 0px 2px 6px rgba(0, 0, 0, 0.25);
        padding: ${toUnits(theme.spacing.padding.large)};
      `}
      {...otherProps}
    >
      {children}
    </div>
  );
};

Card.defaultProps = {
  children: undefined,
};

Card.propTypes = {
  /** @ignore */
  children: PropTypes.node,
};

export const { defaultProps, propTypes } = Card;

export default Card;
