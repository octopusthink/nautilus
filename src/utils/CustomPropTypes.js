import { Children } from 'react';

const allowedChildren = (...types) => {
  return (propValue, key, componentName) => {
    let error = null;

    Children.forEach(propValue[key], (child) => {
      if (!types.includes(child.type)) {
        error = new Error(
          `${componentName} component only accepts the following components as children: ${types
            .map((element) => element.displayName || element)
            .join(', ')}.`,
        );
      }
    });

    return error;
  };
};

export default {
  allowedChildren,
};
