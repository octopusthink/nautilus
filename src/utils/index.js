export { default as CustomPropTypes } from './CustomPropTypes';
