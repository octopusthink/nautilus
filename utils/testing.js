// We only use react-testing-library in development, so ignore this linting
// error.
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { createMemoryHistory } from 'history';
import PropTypes from 'prop-types';
import React from 'react';
import { Router } from 'react-router-dom';

import { Nautilus } from 'components';
import { nautilus as nautilusDefaultTheme, themePropTypes } from 'themes';

// Wrap `react-testing-library`'s `render` function with our providers.
const NautilusProviders = ({ children, theme }) => {
  return <Nautilus theme={theme}>{children}</Nautilus>;
};

NautilusProviders.defaultProps = {
  children: undefined,
  theme: nautilusDefaultTheme,
};

NautilusProviders.propTypes = {
  /** @ignore */
  children: PropTypes.node,
  /** @ignore */
  theme: themePropTypes,
};

const customRender = (ui, options) => {
  return render(ui, { wrapper: NautilusProviders, ...options });
};

// this is a handy function that I would utilize for any component
// that relies on the router being in context
export const renderWithRouter = (
  ui,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
  } = {},
) => {
  return {
    ...customRender(<Router history={history}>{ui}</Router>),
    // adding `history` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    history,
  };
};

export const muteConsole = ({ times, type } = {}) => {
  Array.from({ length: times }).forEach(() => {
    global.console[type].mockImplementationOnce(() => jest.fn());
  });
};

// Export jest-axe helpers.
export { axe } from 'jest-axe';

// Re-export everything else as normal.
export * from '@testing-library/react';

// override render method
export { customRender as render };
