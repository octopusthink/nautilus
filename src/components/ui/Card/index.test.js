import React from 'react';

import { axe, render } from 'utils/testing';

import Card from '.';

describe('Card', () => {
  it('should render a <div>', () => {
    const { container } = render(<Card>My Account</Card>);

    expect(container.firstChild.tagName).toEqual('DIV');
  });

  it('should output its children', () => {
    const { getByTestId } = render(
      <Card>
        <div data-testid="child" />
      </Card>,
    );

    expect(getByTestId('child')).toBeDefined();
  });

  it('should output card CSS', () => {
    const { container } = render(<Card>My Blog Posts</Card>);

    expect(container.firstChild).toMatchSnapshot();
  });

  describe('accessibility', () => {
    it('should pass aXe tests', async () => {
      const { container } = render(<Card>My Blog Posts</Card>);

      expect(await axe(container.innerHTML)).toHaveNoViolations();
    });
  });
});
