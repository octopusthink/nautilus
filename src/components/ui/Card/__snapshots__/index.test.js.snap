// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should output card CSS 1`] = `
.emotion-0 {
  background: white;
  border: 1px solid #f8f9fa;
  box-shadow: 0px 4px 8px rgba(0,0,0,0.125);
  padding: 2.4rem;
  margin-bottom: 3.2rem;
}

.emotion-0 .Nautilus-Heading-1,
.emotion-0 .Nautilus-Heading-2,
.emotion-0 .Nautilus-Heading-3,
.emotion-0 .Nautilus-Heading-4,
.emotion-0 .Nautilus-Heading-5,
.emotion-0 .Nautilus-Heading-6 {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 600;
  font-size: 2.7rem;
  line-height: 1.3333333333333333;
  color: #666c76;
}

.emotion-0 p {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 400;
  font-size: 1.7rem;
  line-height: 1.6470588235294117;
}

.emotion-0 button {
  font-family: -apple-system,BlinkMacSystemFont,San Francisco,Roboto,Segoe UI,Helvetica Neue,sans-serif;
  font-weight: 500;
  font-size: 1.7rem;
  line-height: 1.411764705882353;
  padding: 0.8rem 1.2rem;
}

<div
  class="emotion-0"
>
  My Blog Posts
</div>
`;
